internal/pkg/oras/oras.go

```diff
@@ -0,0 +1,183 @@
+	handlerFunc := func(ctx context.Context, desc ocispec.Descriptor) ([]ocispec.Descriptor, error) {
+		if desc.MediaType == SifLayerMediaType {
+			// Ensure descriptor is of a single file
+			// AnnotationUnpack indicates that the descriptor is of a directory
+			if desc.Annotations[content.AnnotationUnpack] == "true" {
+				return nil, fmt.Errorf("descriptor is of a bundled directory, not a SIF image")
+			}
+			nameOld, _ := content.ResolveName(desc)
+			_ = store.MapPath(nameOld, imagePath)
+		}
+		return nil, nil
+	}
+	pullHandler := oras.WithPullBaseHandler(images.HandlerFunc(handlerFunc))
+
+	_, _, err = oras.Pull(orasctx.Background(), resolver, spec.String(), store, allowedMediaTypes, pullHandler)
+	if err != nil {
+		return fmt.Errorf("unable to pull from registry: %s", err)
+	}
+
+	// ensure that we have downloaded a SIF
+	if err := EnsureSIF(imagePath); err != nil {
+		// remove whatever we downloaded if it is not a SIF
+		os.RemoveAll(imagePath)
+		return err
+	}
+
+	// ensure container is executable
+	if err := os.Chmod(imagePath, 0755); err != nil {
+		return fmt.Errorf("unable to set image perms: %s", err)
+	}
+
+	return nil
+}
+
+// EnsureSIF checks for a SIF image at filepath and returns an error if it is not, or an error is encountered
+func EnsureSIF(filepath string) error {
+	img, err := image.Init(filepath, false)
+	if err != nil {
+		return fmt.Errorf("could not open image %s for verification: %s", filepath, err)
+	}
+	defer img.File.Close()
+
+	if img.Type != image.SIF {
+		return fmt.Errorf("%q is not a SIF", filepath)
+	}
+
+	return nil
+}
+
+// ImageSHA returns the sha256 digest of the SIF layer of the OCI manifest
+// oci spec dictates only sha256 and sha512 are supported at time creation for this function
+// sha512 is currently optional for implementations, this function will return an error when
+// encountering such digests.
+// https://github.com/opencontainers/image-spec/blob/master/descriptor.md#registered-algorithms
+func ImageSHA(uri string, sysCtx *ocitypes.SystemContext) (string, error) {
+	imgref, err := dockerref.ParseReference(uri)
+	if err != nil {
+		return "", fmt.Errorf("while parsing reference: %v", err)
+	}
+
+	img, err := imgref.NewImage(context.Background(), sysCtx)
+	if err != nil {
+		return "", fmt.Errorf("while creating new image descriptor: %v", err)
+	}
+	defer img.Close()
+
+	// search image layers for sif image and return sha
+	for _, l := range img.LayerInfos() {
+		if l.MediaType == SifLayerMediaType {
+			// only allow sha256 digests
+			if l.Digest.Algorithm() != digest.SHA256 {
+				return "", fmt.Errorf("SIF layer found with incorrect digest algorithm: %s", l.Digest.Algorithm())
+			}
+			return l.Digest.String(), nil
+		}
+	}
+
+	return "", fmt.Errorf("no layer found corresponding to SIF image")
+}
+
+// ImageHash returns the appropriate hash for a provided image file
+//   e.g. sha256:<sha256>
+func ImageHash(filePath string) (result string, err error) {
```

Comment:
We start to duplicate what was in the client package for library images and that is now in https://github.com/sylabs/scs-library-client/blob/cec4defe222a191539b6a777c33df76bb33637bb/client/util.go#L131
Is it okay? I just do not know...

Reply:
This is very similar but there are two caveats that creating a function here addresses.
1. the `scs-library-client` gives a sha with a `.` separator when I want a `:` to match the format of OCI layer digests.
2. That client function has a note to change to returning the sif uuid in the future which I want to ensure does not happen here since detecting identical images is reliant on the `sha` sum due to using the OCI layer digest to identify unique images